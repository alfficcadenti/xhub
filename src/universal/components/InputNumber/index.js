import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {FormInput} from '@homeaway/react-form-components';
import h from '../../components/utils/formatString'

class InputNumber extends Component {
    constructor(props) {
        super(props);
        this.state = {
            value: '',
            error: ''
        }
    }

    onlyInteger = (event) => (event.key==='.' ? true : false)

    onlyPositives = (event) => (event.key==='-' ? true : false)

    outsideRange = (value,min,max) => ((!(max && value > max) && !(min && value < min)) ? false : true)

    decimals = (value) => {
        if(Math.floor(value) === value) return 0;
        return value.toString().split(".")[1] || 0; 
    }

    helpTextString = (range) => {
        if (range.min && !range.max) {
            return `min value ${range.min}`
        }
        if (range.min && range.max) {
            return `value between ${range.min} - ${range.max}`
        }
    }

    validation = (event) => {
        const {
            range,
            type,
        } = this.props;
        if(type === 'integer') {
            if(this.onlyInteger(event)) {event.preventDefault()}
        } else {
            if(this.decimals(event.target.value) && this.decimals(event.target.value).length >= 2) {event.preventDefault();}
        }
        if(range.min >= 0) {
            if(this.onlyPositives(event)) {event.preventDefault()}
        }
    }

    checkValue = (event) => {
        const {range} = this.props;

        if(this.outsideRange(event.target.value, Number(range.min), Number(range.max))) {
            this.setState({
                error: `Error: value is not valid`,
                value: event.target.value
            })
        } else {
            this.setState({
                value: event.target.value, 
                error:''
            })
        }
    }

    render() {
        const {
            id,
            question,
            range
        } = this.props;

        const helpText = (range.min || range.max) ? this.helpTextString(range) : ''

        return (
            <FormInput
                key={id}
                id={h.replaceSpaces(question)}
                label={question}
                type='number'
                onChange={this.checkValue}
                onKeyPress={this.validation}
                value={this.state.value}
                errorMsg={this.state.error}
                helpText={helpText}
            />
        )
    }
}

InputNumber.defaultProps = {
    id: null,
    question: '',
    type: null,
    range: {min: '', max:''}
};

InputNumber.propTypes = {
    id: PropTypes.string,
    question: PropTypes.string,
    type: PropTypes.string,
    range: PropTypes.object,
};

export default InputNumber;