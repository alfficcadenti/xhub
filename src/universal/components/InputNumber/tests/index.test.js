import React from 'react';
import {mount} from 'enzyme';
import chai, {expect} from 'chai';
import chaiJestSnapshot from 'chai-jest-snapshot';
import InputNumber from '../index';
import renderer from 'react-test-renderer';
chai.use(chaiJestSnapshot);

const questions = [
    {id: 0, question: '# Availability Zones Deployed to', type: 'integer'},
    {id: 1, question: '% Prod Traffic', type: 'integer', range: {min: 0, max: 100}}
];

describe('<InputNumber /> ', () => {
    beforeEach(() => {
        chaiJestSnapshot.setFilename(`${__filename}.snap`);
    });

    it('renders correctly', () => {
        const props = questions[1];
        const wrapper = mount(<InputNumber {...props} />);
        expect(wrapper).to.have.lengthOf(1);
    });

    it('matches the snapshot', () => {
        const props = questions[1];
        chaiJestSnapshot.setTestName('matches the snapshot');
        const wrapper = renderer.create(<InputNumber {...props} />);
        expect(wrapper).to.matchSnapshot();
    });

    it('renders a span.help-block when a value is outside the range', () => {
        const props = questions[1];
        const wrapper = mount(<InputNumber {...props} />);
        const input = wrapper.find('input');
        input.simulate('change', {target: {value: '5000'}});
        expect(wrapper.find('span.help-block').text()).to.includes('Error');
    });

    describe('onlyInteger()', () => {
        it('return true when event.key is a dot', () => {
            const event = {key: '.'};
            const props = questions[1];
            const wrapper = mount(<InputNumber {...props} />);
            expect(wrapper.instance().onlyInteger(event)).to.be.true;
        });

        it('return false when event.key is not a dot', () => {
            const event = {key: 'a'};
            const props = questions[1];
            const wrapper = mount(<InputNumber {...props} />);
            expect(wrapper.instance().onlyInteger(event)).to.be.false;
        });
    });

    describe('onlyPositives()', () => {
        it('return true when event.key is a -', () => {
            const event = {key: '-'};
            const props = questions[1];
            const wrapper = mount(<InputNumber {...props} />);
            expect(wrapper.instance().onlyPositives(event)).to.be.true;
        });

        it('return false when event.key is not a -', () => {
            const event = {key: 'a'};
            const props = questions[1];
            const wrapper = mount(<InputNumber {...props} />);
            expect(wrapper.instance().onlyPositives(event)).to.be.false;
        });
    });

    describe('outsideRange()', () => {
        it('return true when the value in input is within the range of min max in input', () => {
            const value = 555;
            const min = 0;
            const max = 100;
            const props = questions[1];
            const wrapper = mount(<InputNumber {...props} />);
            expect(wrapper.instance().outsideRange(value, min, max)).to.be.true;
        });

        it('return true when the value in input is within the range of min max in input', () => {
            const value = 0;
            const min = 1;
            const max = 100;
            const props = questions[1];
            const wrapper = mount(<InputNumber {...props} />);
            expect(wrapper.instance().outsideRange(value, min, max)).to.be.true;
        });

        it('return false when the value in input is outside the range of min max in input', () => {
            const value = 55;
            const min = 0;
            const max = 100;
            const props = questions[1];
            const wrapper = mount(<InputNumber {...props} />);
            expect(wrapper.instance().outsideRange(value, min, max)).to.be.false;
        });
    });

    describe('decimals()', () => {
        it('return a string with the decimals of a number in input', () => {
            const value = 555.77;
            const props = questions[1];
            const wrapper = mount(<InputNumber {...props} />);
            expect(wrapper.instance().decimals(value)).to.be.eql('77');
        });

        it('return 0 when hte number in input has no decimals', () => {
            const value = 555;
            const props = questions[1];
            const wrapper = mount(<InputNumber {...props} />);
            expect(wrapper.instance().decimals(value)).to.be.eql(0);
        });
    });
});