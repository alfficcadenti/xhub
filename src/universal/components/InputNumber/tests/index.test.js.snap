// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div
  className="form-group floating-label empty"
>
  <input
    aria-invalid={false}
    aria-label="% Prod Traffic"
    className="form-control"
    id="%-Prod-Traffic"
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="% Prod Traffic"
    readOnly={undefined}
    type="number"
    value=""
  />
  <label
    className=""
    htmlFor="%-Prod-Traffic"
  >
    % Prod Traffic
  </label>
  
  
</div>
`;
